src/lang/encoders.ml

```diff
@@ -0,0 +1,212 @@
+          List.exists
+            (function "", `Encoder ("theora", _) -> true | _ -> false)
+            p
+        in
+        if not video then Encoder.audio_kind channels
+        else Encoder.audio_video_kind channels
+    | "ffmpeg" ->
+        let audio =
+          List.fold_left
+            (fun audio p ->
+              match p with
+                | "", `Encoder ("audio.copy", _) ->
+                    `Format
+                      Frame_content.(
+                        default_format (kind_of_string "ffmpeg.audio.copy"))
+                | "", `Encoder ("audio.raw", _) ->
+                    `Format
+                      Frame_content.(
+                        default_format (kind_of_string "ffmpeg.audio.raw"))
+                | "", `Encoder ("audio", p) ->
+                    let channels =
+                      try
+                        let channels =
+                          try List.assoc "channels" p
+                          with Not_found -> List.assoc "ac" p
+                        in
+                        match channels with
+                          | `Term { term = Ground (Int n) } -> n
+                          | _ -> raise Exit
+                      with
+                        | Not_found -> 2
+                        | Exit -> raise Not_found
+                    in
+                    `Format
+                      Frame_content.(
+                        Audio.lift_params
+                          {
+                            Contents.channel_layout =
+                              lazy
+                                (Audio_converter.Channel_layout
+                                 .layout_of_channels channels);
+                          })
+                | _ -> audio)
+            Frame.none p
+        in
+        let video =
+          List.fold_left
+            (fun audio p ->
+              match p with
+                | "", `Encoder ("video.copy", _) ->
+                    `Format
+                      Frame_content.(
+                        default_format (kind_of_string "ffmpeg.video.copy"))
+                | "", `Encoder ("video.raw", _) ->
+                    `Format
+                      Frame_content.(
+                        default_format (kind_of_string "ffmpeg.video.raw"))
+                | "", `Encoder ("video", _) ->
+                    `Format Frame_content.(default_format Video.kind)
+                | _ -> audio)
+            Frame.none p
+        in
+        { Frame.audio; video; midi = Frame.none }
+    | _ -> raise Not_found
+
+let type_of_encoder ~pos ~level f =
+  let kind = kind_of_encoder f in
+  let audio = kind_t ~pos ~level kind.Frame.audio in
+  let video = kind_t ~pos ~level kind.Frame.video in
+  let midi = kind_t ~pos ~level kind.Frame.midi in
+  format_t ~pos ~level (frame_kind_t ~pos ~level audio video midi)
+
+exception Error
+
+(* TODO: it would be better if each encoder registered itself here than
+   having this match *)
```

Comment:
Another thing that can be achieved by extensible variant type. You declare a base type for encoders with whatever you need to register them, then each encoder module uses this interface to register itself with all its required bits. This would also work potentially for dynamically loaded encoders (though I'd _really_ like to avoid it).

See for instance:
```ocaml
module type AbstractDef = sig
  type content

  val name : string
  val to_json : compact:bool -> json5:bool -> content -> string
  val descr : content -> string
  val compare : content -> content -> int
end

module MkAbstract (Def : AbstractDef) = struct
  type Type.ground += Type
  type G.t += Value of Def.content

  let () =
    G.register (function
      | Value v ->
          let compare = function
            | Value v' -> Def.compare v v'
            | _ -> assert false
          in
          Some
            {
              G.descr = (fun () -> Def.descr v);
              to_json =
                (fun ~compact ~json5 () -> Def.to_json ~compact ~json5 v);
              compare;
              typ = Type;
            }
      | _ -> None);

...
```

Reply:
This is a bit of work. Would it be ok to do this in a second PR?